From 1c8dcfaf25380c70ea355431510f8445ed461f23 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:37:07 -0400
Subject: [virt] kvm: add kvm_arch_vcpu_runnable() test to kvm_vcpu_on_spin() loop

Message-id: <1406040016-3289-33-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86111
O-Subject: [RHEL7 PATCH v2 032/221] KVM: add kvm_arch_vcpu_runnable() test to kvm_vcpu_on_spin() loop
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Michael Mueller <mimu@linux.vnet.ibm.com>

Use the arch specific function kvm_arch_vcpu_runnable() to add a further
criterium to identify a suitable vcpu to yield to during undirected yield
processing.

Signed-off-by: Michael Mueller <mimu@linux.vnet.ibm.com>
Reviewed-by: Christian Borntraeger <borntraeger@de.ibm.com>
Signed-off-by: Christian Borntraeger <borntraeger@de.ibm.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 98f4a14676127397c54cab7d6119537ed4d113a2)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 80a1e04..29bc835 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -1793,7 +1793,7 @@ void kvm_vcpu_on_spin(struct kvm_vcpu *me)
     continue;
    if (vcpu == me)
     continue;
-   if (waitqueue_active(&vcpu->wq))
+   if (waitqueue_active(&vcpu->wq) && !kvm_arch_vcpu_runnable(vcpu))
     continue;
    if (!kvm_vcpu_eligible_for_directed_yield(vcpu))
     continue;
-- 
1.7.1